tool/src/js/type_generation/mod.rs

```diff
@@ -162,52 +162,59 @@ impl<'ctx, 'tcx> TyGenContext<'ctx, 'tcx> {
-                        panic!("'static not supported in JS2 backend");
+                        panic!("'static not supported in JS backend");
                     };
-                    format!(
-                        "(appendArrayMap[{lt_name}AppendArray] || []).length > 0 ? () => {{ for (let lifetime of appendArrayMap[{lt_name}AppendArray]) {{ appendArrayMap[{lt_name}AppendArray].push({field_name}); }} {field_name}.garbageCollect(); }} : {field_name}.free",
-                        lt_name = struct_def.lifetimes.fmt_lifetime(lt),
-                    )
-                } else {
-                    // We take ownership
-                    "".into()
-                };
-
-                (format!("{slice_expr}"), Some(post_cleanup_statement), None)
-            } else {
-                let borrow_info = if let hir::Type::Struct(path) = &field.ty {
-                    StructBorrowInfo::compute_for_struct_field(struct_def, path, self.tcx).map(
-                        |param_info| StructBorrowContext {
-                            use_env: &struct_def.lifetimes,
-                            param_info,
-                            is_method: false
-                        }
+                    Some(
+                        format!(
+                            r#"(appendArrayMap["{lt_name}AppendArray"].length > 0 ? diplomatRuntime.CleanupArena.createWith(appendArrayMap["{lt_name}AppendArray"]) : functionCleanupArena)"#,
+                            lt_name = struct_def.lifetimes.fmt_lifetime(lt),
+                        )
                     )
                 } else {
                     None
-                };
+                }
+            } else if let &hir::Type::Struct(..) = &field.ty {
+                Some("functionCleanupArena".into())
+            } else {
+                // We take ownership
+                None
+            };
 
-                (self.gen_js_to_c_for_type(&field.ty, format!("this.#{}", field_name.clone()).into(), borrow_info.as_ref()).into(), None, borrow_info.map(|s| s.param_info))
+            let maybe_struct_borrow_info = if let hir::Type::Struct(path) = &field.ty {
+                StructBorrowInfo::compute_for_struct_field(struct_def, path, self.tcx).map(
+                    |param_info| StructBorrowContext {
+                        use_env: &struct_def.lifetimes,
+                        param_info,
+                        is_method: false
+                    }
+                )
+            } else {
+                None
             };
 
+            let js_to_c = format!("{}{}{}", 
+                match &field.ty {
+                    Type::Slice(..) => "...",
+                    _ => "",
+                },
+                self.gen_js_to_c_for_type(&field.ty, format!("this.#{}", field_name.clone()).into(), maybe_struct_borrow_info.as_ref(), alloc.as_deref()),
+                match &field.ty {
+                    Type::Slice(..) => ".splat()",
```

Comment:
issue: I would like for the splatting/etc to be handled by gen_js_to_c, it would be nice if that function handled all of the special casing so it can be reused in other code.

For example for Option we're going to need to invoke it again and it feels strange to special case it. At the moment gen_js_to_c returns the right thing _except_ for slices, which seems off.